src/client/dfuse/pil4dfs/dfs_dcache.c

```diff
@@ -0,0 +1,616 @@
+{
+	dfs_dcache_t *dcache_tmp = NULL;
+	dfs_obj_t    *obj        = NULL;
+	daos_obj_id_t obj_id;
+	mode_t        mode;
+	int           rc;
+
+	D_ALLOC_PTR(dcache_tmp);
+	if (dcache_tmp == NULL)
+		D_GOTO(error, rc = -DER_NOMEM);
+
+	dcache_tmp->dd_dfs = dfs;
+
+	rc = dfs_lookup(dfs, "/", O_RDONLY, &obj, &mode, NULL);
+	if (rc != 0)
+		D_GOTO(error, rc = daos_der2errno(rc));
+	rc = dfs_obj2id(obj, &obj_id);
+	D_ASSERT(rc == 0);
+	rc = snprintf(&dcache_tmp->dd_key_root_prefix[0], DCACHE_KEY_PREF_SIZE,
+		      "%016" PRIx64 "-%016" PRIx64 ":", obj_id.hi, obj_id.lo);
+	D_ASSERT(rc == DCACHE_KEY_PREF_SIZE - 1);
+
+	rc = d_hash_table_create_inplace(D_HASH_FT_MUTEX | D_HASH_FT_LRU, bits, NULL,
+					 &dcache_hash_ops, &dcache_tmp->dd_dir_hash);
+	if (rc != 0)
+		D_GOTO(error, rc);
+
+	dcache_tmp->destroy_fn     = dcache_destroy_act;
+	dcache_tmp->find_insert_fn = dcache_find_insert_act;
+	dcache_tmp->drec_incref_fn = drec_incref_act;
+	dcache_tmp->drec_decref_fn = drec_decref_act;
+	dcache_tmp->drec_del_at_fn = drec_del_at_act;
+	dcache_tmp->drec_del_fn    = drec_del_act;
+
+	*dcache = dcache_tmp;
+	D_GOTO(out, rc = -DER_SUCCESS);
+
+error:
+	D_FREE(dcache_tmp);
+out:
+	dfs_release(obj);
+	return rc;
+}
+
+static int
+dcache_destroy_act(dfs_dcache_t *dcache)
+{
+	d_list_t *rlink = NULL;
+	int       rc;
+
+	while ((rlink = d_hash_rec_first(&dcache->dd_dir_hash)) != NULL) {
+		bool deleted;
+
+		D_ASSERT(atomic_load(&dlist2drec(rlink)->dr_ref) == 1);
+		deleted = d_hash_rec_delete_at(&dcache->dd_dir_hash, rlink);
+		D_ASSERT(deleted);
+	}
+
+	rc = d_hash_table_destroy_inplace(&dcache->dd_dir_hash, false);
+	if (rc != 0) {
+		DL_ERROR(rc, "d_hash_table_destroy_inplace() failed");
+		D_GOTO(out, rc);
+	}
+
+	D_FREE(dcache);
+
+out:
+	return rc;
+}
+
+static inline dcache_rec_t *
+dcache_get(dfs_dcache_t *dcache, const char *key, size_t key_len)
+{
+	d_list_t *rlink;
+
+	D_ASSERT(dcache != NULL);
+	D_ASSERT(key != NULL);
+	D_ASSERT(key_len > 0);
+
+	rlink = d_hash_rec_find(&dcache->dd_dir_hash, key, key_len);
+	if (rlink == NULL)
+		return NULL;
+	return dlist2drec(rlink);
+}
+
+static inline int
+dcache_add(dcache_rec_t **rec, dfs_dcache_t *dcache, const char *path, const char *key,
+	   size_t key_len)
+{
+	dcache_rec_t *rec_tmp = NULL;
+	dfs_obj_t    *obj     = NULL;
+	daos_obj_id_t obj_id;
+	mode_t        mode;
+	d_list_t     *rlink;
+	int           rc;
+
+	D_ALLOC(rec_tmp, sizeof(*rec_tmp) + key_len + 1);
+	if (rec_tmp == NULL)
+		D_GOTO(error, rc = -DER_NOMEM);
+
+	atomic_init(&rec_tmp->dr_ref, 1);
+	atomic_init(&rec_tmp->dr_deleted, false);
+
+	rc = dfs_lookup(dcache->dd_dfs, path, O_RDWR, &obj, &mode, NULL);
+	if (rc != 0)
+		D_GOTO(error, rc = d_errno2der(rc));
+	if (!S_ISDIR(mode))
+		D_GOTO(error, rc = -DER_NOTDIR);
+	rec_tmp->dr_obj = obj;
+
+	rc = dfs_obj2id(obj, &obj_id);
+	D_ASSERT(rc == 0);
+	rc = snprintf(&rec_tmp->dr_key_child_prefix[0], DCACHE_KEY_PREF_SIZE,
+		      "%016" PRIx64 "-%016" PRIx64 ":", obj_id.hi, obj_id.lo);
```

Comment:
Text version of obj_id is used here. Shall we use the original 16 bytes obj_id instead to make the key shorter?

Reply:
From my side, I would prefer to keep the textual version: makes things easier for maintenance and debugging.  I am also not very found of mixing binary values (i.e. obj_id) and textual values (file name).